sdk/storage/Azure.Storage.Common/src/Shared/StorageConnectionString.cs

```diff
@@ -545,22 +446,34 @@ string settingOrDefault(string key)
                 var queueSecondaryEndpoint = settingOrDefault(Constants.ConnectionStrings.QueueSecondaryEndpointSetting);
                 var tableSecondaryEndpoint = settingOrDefault(Constants.ConnectionStrings.TableSecondaryEndpointSetting);
                 var fileSecondaryEndpoint = settingOrDefault(Constants.ConnectionStrings.FileSecondaryEndpointSetting);
+                var sasToken = settingOrDefault(Constants.ConnectionStrings.SharedAccessSignatureSetting);
 
                 // if secondary is specified, primary must also be specified
 
                 static bool s_isValidEndpointPair(string primary, string secondary) =>
                         !string.IsNullOrWhiteSpace(primary)
                         || /* primary is null, and... */ string.IsNullOrWhiteSpace(secondary);
 
-                (Uri, Uri) createStorageUri(string primary, string secondary, Func<IDictionary<string, string>, (Uri, Uri)> factory) =>
+                (Uri, Uri) createStorageUri(string primary, string secondary, string sasToken, Func<IDictionary<string, string>, (Uri, Uri)> factory)
+                {
+                    return
                         !string.IsNullOrWhiteSpace(secondary) && !string.IsNullOrWhiteSpace(primary)
-                            ? (new Uri(primary), new Uri(secondary))
+                            ? (CreateUri(primary, sasToken), CreateUri(secondary, sasToken))
                         : !string.IsNullOrWhiteSpace(primary)
-                            ? (new Uri(primary), default)
+                            ? (CreateUri(primary, sasToken), default)
                         : matchesAutomaticEndpointsSpec && factory != null
                             ? factory(settings)
-                        : (default, default)
-                        ;
+                        : (default, default);
+
+                    static Uri CreateUri(string endpoint, string sasToken)
+                    {
+                        var builder = new UriBuilder(endpoint)
+                        {
+                            Query = sasToken
```

Comment:
Do we know for sure the endpoint will never have any query params?  Especially since connection strings aren't well defined anywhere?

Reply:
Would you recommend using HttpUtility.ParseQueryString to append sasToken?

Reply:
No, I don't think we need to parse it - I just don't want to stomp any values.  I'd create the builder first without setting `Query`.  If `sasToken` isn't null/empty, then we can check whether `builder.Query` has a value before appending (and possibly squeezing in a `&`).

Reply:
Updated